File: native/frontend/src/contract.ts
/**
 * Entropy sidecar wire contract.
 *
 * This module is the single typed source of truth for everything that crosses
 * the HTTP/WebSocket boundary. Schema version 2.
 */

export const SCHEMA_VERSION = 2

/* ------------------------------------------------------------------ *
 * Live snapshot (ws://127.0.0.1:{port}/ws/live, ~10 Hz)
 * ------------------------------------------------------------------ */

export type Leader = [symbol: string, count: number, price: number, pct: number]
export type TickerEntry = [window: string, entries: [string, number][]]
export type Candle = [tsNs: number, o: number, h: number, l: number, c: number, v: number]
export type Level = [price: number, size: number]
// The canonical id stays first because everything keys off it, but the row now
// also carries what a human should actually read: the exchange's own ticker,
// the instrument name and the venue badge. Rendering the canonical id alone
// clipped every crypto row to "binance-s…".
export type WatchRow = [
  symbol: string,
  ticker: string,
  name: string,
  exchange: string,
  last: number | null,
  pct: number | null,
  spark: number[],
]

export type AssetClass = 'EQUITY' | 'CRYPTO' | 'SIM'

/** Feed lifecycle states reported per venue class. */
export type FeedState = 'off' | 'sim' | 'live' | 'connecting' | 'error'

export interface DepthLevels {
  basis: string
  is_synthetic: boolean
  reference_price: number
  bids: Level[]
  asks: Level[]
}

export interface Fundamentals {
  pe: number | null
  market_cap: number | null
  high_52w: number | null
  low_52w: number | null
}

export interface FocusView {
  symbol: string
  asset: AssetClass
  last: number | null
  pct: number | null
  hi: number | null
  lo: number | null
  candles: Candle[]
  depth: DepthLevels | null
  fundamentals: Fundamentals | null
  /** Resolved chart interval actually used to bucket `candles`, e.g. "1m". */
  interval: string
  /** Scanner timeframe currently driving the high/low windows, e.g. "15m". */
  timeframe: string
}

/** The nine settings echoed on every frame so the UI never has to guess. */
export interface SettingsView {
  timeframe: string
  chart_interval: string
  chart_type: string
  show_volume: boolean
  show_depth: boolean
  equity_source: string
  enable_equities: boolean
  enable_crypto: boolean
  theme: string
}

export interface FeedStatus {
  equities: string
  crypto: string
  detail: string
}

export interface BotPosition {
  symbol: string
  side: string
  qty: number
  entry_px: number
  mark_px: number
  unrealized_pnl: number
  stop_px: number
  tp_px: number
}

export interface BotStrategy {
  name: string
  warm: boolean
  /** symbol -> regime label ("trend" | "range" | "chop" | …) */
  regimes: Record<string, string>
  /** symbol -> directional vote in {-1, 0, +1} */
  directions: Record<string, number>
}

export interface BotView {
  running: boolean
  paused: boolean
  halted: boolean
  warm: boolean
  mode: string
  timeframe: string
  bar_s: number
  risk_profile: string
  risk_description: string
  ticks: number
  cash: number
  equity: number
  realized_pnl: number
  unrealized_pnl: number
  daily_pnl: number
  open_count: number
  positions: BotPosition[]
  strategies: BotStrategy[]
  last_signals: string[]
  last_rejects: string[]
}

export interface SnapshotMessage {
  type: 'snapshot'
  schema_version: number
  ts_ns: number
  buy_pct: number
  sell_pct: number
  raw_hz: number
  accel: string
  new_highs: Leader[]
  new_lows: Leader[]
  ticker: TickerEntry[]
  focus: FocusView
  watchlist: WatchRow[]
  market_status: string
  source: string
  settings: SettingsView
  feeds: FeedStatus
  bot: BotView | null
}

/* ------------------------------------------------------------------ *
 * REST
 * ------------------------------------------------------------------ */

/** Uniform mutation acknowledgement. `problems` is human-readable. */
export interface ApiAck {
  ok: boolean
  message: string
  problems: string[]
}

export interface RiskProfileMeta {
  name: string
  description: string
  /** Textual colour name supplied by the sidecar; advisory only. */
  color?: string
  per_trade_pct: number
  max_concurrent: number
  stop_loss_pct: number
  take_profit_pct: number
  max_total_exposure_pct: number
  max_daily_loss_pct: number
  cooldown_s: number
  min_volatility_pct: number
  vol_window_s: number
}

export interface MetaResponse {
  timeframes: string[]
  chart_intervals: string[]
  themes: string[]
  strategies: string[]
  risk_profiles: RiskProfileMeta[]
  vote_modes: string[]
  normalize_modes: string[]
  exit_modes: string[]
  equity_sources: string[]
  settings_path: string
}

export interface SymbolRow {
  /** Canonical feed id — "AAPL", "binance-spot:BTCUSDT". Never shown raw. */
  symbol: string
  /** What the instrument IS: "Apple Inc.", "Bitcoin / TetherUS". */
  name: string
  asset_class: string
  venue: string
  /** The exchange's own symbol: "AAPL", "BTCUSDT", "BTC-USD". */
  ticker: string
  /** Badge text: "US", "BINANCE", "COINBASE". */
  exchange: string
  /** Crypto only. */
  base: string
  quote: string
  watched: boolean
}

export interface WatchlistEntry {
  symbol: string
  name: string
  asset_class: string
  venue: string
  ticker: string
  exchange: string
  base: string
  quote: string
}

/* ---- persisted configuration ------------------------------------- */

export interface AppConfigJSON {
  seed: number
  equity_tps: number
  enable_crypto: boolean
  enable_equities: boolean
  equity_source: string
  strategy_symbol: string
  crypto_strategy_symbol: string
  theme: string
  chart_type: string
  show_volume: boolean
  show_depth: boolean
  depth_bins: number
  depth_top_n: number
  risk_profile: string
  console_log_path: string
  trade_csv_path: string
  timeframe: string
  /** "" means: follow the scanner timeframe. */
  chart_interval: string
  chart_bars: number
  watchlist_path: string
  /**
   * Engine sub-config owned by the sidecar. Read-only here: the PUT deep-merges,
   * so omitting the key preserves it, and sending a partial would be rejected.
   */
  engine?: Record<string, unknown>
}

export interface RiskOverridesJSON {
  per_trade_pct: number | null
  max_concurrent: number | null
  stop_loss_pct: number | null
  take_profit_pct: number | null
  max_total_exposure_pct: number | null
  max_daily_loss_pct: number | null
  cooldown_s: number | null
  min_volatility_pct: number | null
  vol_window_s: number | null
}

export interface ConsensusConfigJSON {
  threshold: number
  min_bars: number
  vote_mode: string
  normalize: string
  min_participation: number
  w_ema: number
  w_macd: number
  w_rsi: number
  w_bollinger: number
  ema_fast: number
  ema_slow: number
  macd_fast: number
  macd_slow: number
  macd_signal: number
  rsi_period: number
  rsi_low: number
  rsi_high: number
  rsi_trend_low: number
  rsi_trend_high: number
  bb_period: number
  bb_std: number
  bb_low: number
  bb_high: number
  bb_trend_low: number
  bb_trend_high: number
  move_floor: number
  trend_er: number
  regime_window: number
  slope_lookback: number
  regime_tilt: number
  min_hold_bars: number
  cooldown_bars: number
  exit_mode: string
}

export interface BotLiveJSON {
  enabled: boolean
  acknowledged_risk: boolean
  exchange: string
  api_key: string
  api_secret: string
}

export interface BotConfigJSON {
  mode: string
  risk_profile: string
  risk_overrides: RiskOverridesJSON
  strategies: string[]
  symbols: string[]
  starting_cash: number
  fee_bps: number
  slippage_bps: number
  ema_symbol: string
  ema_fast: number
  ema_slow: number
  momentum_min_pct: number
  seed: number
  equity_tps: number
  enable_crypto: boolean
  enable_equities: boolean
  console_log_path: string
  trade_csv_path: string
  live: BotLiveJSON
  timeframe: string
  /** 0 = follow `timeframe`; otherwise an explicit bar width in seconds. */
  bar_s: number
  /** Python bool. Sending a number here is rejected by the sidecar. */
  warmup: boolean
  consensus: ConsensusConfigJSON
}

export interface SettingsPayload {
  app: AppConfigJSON
  bot: BotConfigJSON
}

/**
 * `PUT /api/settings` deep-merges, so nested partials are legal — but unknown
 * keys are rejected outright rather than ignored, and validation is all-or-
 * nothing across both halves.
 */
export interface SettingsPatch {
  app?: Partial<Omit<AppConfigJSON, 'engine'>>
  bot?: Partial<Omit<BotConfigJSON, 'consensus' | 'risk_overrides' | 'live'>> & {
    consensus?: Partial<ConsensusConfigJSON>
    risk_overrides?: Partial<RiskOverridesJSON>
    live?: Partial<BotLiveJSON>
  }
}

export type BotAction = 'start' | 'stop' | 'pause' | 'resume' | 'halt'

File: native/sidecar/entropy_sidecar/contract.py
"""Wire types shared with the React frontend.

Every message the sidecar emits and every body it accepts is declared here as a
msgspec Struct, so the schema is single-sourced instead of being reconstructed
by hand on either side. FastAPI cannot use a msgspec.Struct as a body model, so
the routes decode request bodies with ``msgspec.json.decode(..., type=X)``.
"""

from __future__ import annotations

from typing import Any

import msgspec

Leader = tuple[str, int, float, float]      # (symbol, count, price, pct)
TickerEntry = tuple[str, list[tuple[str, int]]]  # (window, [(symbol, count)])
Candle = tuple[int, float, float, float, float, float]  # (ts_ns, o, h, l, c, v)
Level = tuple[float, float]                 # (price, size)
#: (symbol, last, pct, spark) — spark is the raw price ring, the frontend renders it.
# (canonical symbol, exchange ticker, instrument name, exchange badge,
#  last, pct, spark). The last three used to be all a row carried, which left
# the pane rendering "binance-s…" — the canonical id clipped to the venue.
WatchRow = tuple[str, str, str, str, float | None, float | None, list[float]]

#: Theme vocabulary, mirroring entropy.ui.theme.ALL_THEMES. Duplicated as plain
#: strings rather than imported: /api/meta must answer without pulling Textual's
#: theme objects into the sidecar's import graph just to read seven names.
THEMES: tuple[str, ...] = (
    "entropy", "dracula", "cyberpunk", "nord", "forest", "monochrome", "sweet",
)

#: Chart renderers the frontend implements (mirrors PriceChart.chart_type).
CHART_TYPES: tuple[str, ...] = ("candlestick", "line")

#: AppConfig.equity_source vocabulary.
EQUITY_SOURCES: tuple[str, ...] = ("sim", "live", "auto")


class DepthLevels(msgspec.Struct, frozen=True):
    basis: str
    is_synthetic: bool
    reference_price: float
    bids: list[Level]
    asks: list[Level]


class Fundamentals(msgspec.Struct, frozen=True):
    pe: float | None = None
    market_cap: float | None = None
    high_52w: float | None = None
    low_52w: float | None = None


class FocusView(msgspec.Struct, frozen=True):
    symbol: str
    asset: str                              # "EQUITY" | "CRYPTO" | "SIM"
    last: float | None
    pct: float | None
    hi: float | None
    lo: float | None
    candles: list[Candle]
    depth: DepthLevels | None
    fundamentals: Fundamentals | None
    # Resolved chart interval (NOT necessarily the timeframe: chart_interval ""
    # follows it, anything else overrides it) plus the scanner's timeframe, so
    # the chart pane can label both without re-deriving the resolution rule.
    interval: str = ""
    timeframe: str = ""


class SettingsView(msgspec.Struct, frozen=True):
    """The AppConfig fields the live panes react to, echoed on every snapshot.

    The frontend reads its current state from the stream rather than re-fetching
    /api/settings after each mutation, so a change made from the command bar or
    a second window shows up everywhere within one tick.
    """

    timeframe: str = ""
    chart_interval: str = ""
    chart_type: str = "candlestick"
    show_volume: bool = True
    show_depth: bool = False
    equity_source: str = "auto"
    enable_equities: bool = True
    enable_crypto: bool = True
    theme: str = "entropy"


class FeedStatus(msgspec.Struct, frozen=True):
    """What the two feeds are actually doing right now.

    ``detail`` carries the last non-fatal explanation (live-feed fallback,
    provider name, a bot exception swallowed on the drain) so the UI can show
    WHY a feed is degraded instead of only that it is.
    """

    equities: str = "off"    # "off" | "sim" | "live" | "error"
    crypto: str = "off"      # "off" | "connecting" | "live" | "error"
    detail: str = ""


class BotPosition(msgspec.Struct, frozen=True):
    symbol: str
    side: str
    qty: float
    entry_px: float
    mark_px: float
    unrealized_pnl: float
    stop_px: float
    tp_px: float


class BotStrategy(msgspec.Struct, frozen=True):
    name: str
    warm: bool
    regimes: dict[str, str] = msgspec.field(default_factory=dict)
    directions: dict[str, int] = msgspec.field(default_factory=dict)


class BotView(msgspec.Struct, frozen=True):
    running: bool
    paused: bool
    halted: bool
    warm: bool
    mode: str
    timeframe: str
    bar_s: float
    risk_profile: str
    risk_description: str
    ticks: int
    cash: float
    equity: float
    realized_pnl: float
    unrealized_pnl: float
    daily_pnl: float
    open_count: int
    positions: list[BotPosition] = msgspec.field(default_factory=list)
    strategies: list[BotStrategy] = msgspec.field(default_factory=list)
    last_signals: list[str] = msgspec.field(default_factory=list)
    last_rejects: list[str] = msgspec.field(default_factory=list)


class SnapshotMessage(msgspec.Struct, frozen=True, tag="snapshot", tag_field="type"):
    schema_version: int
    ts_ns: int
    buy_pct: float
    sell_pct: float
    raw_hz: float
    accel: str
    new_highs: list[Leader]
    new_lows: list[Leader]
    ticker: list[TickerEntry]
    focus: FocusView
    watchlist: list[WatchRow]
    market_status: str
    source: str
    settings: SettingsView = msgspec.field(default_factory=SettingsView)
    feeds: FeedStatus = msgspec.field(default_factory=FeedStatus)
    #: None until the bot has been started once — "never started" and "started
    #: then stopped" are different states and the dashboard shows them so.
    bot: BotView | None = None


class CommandRequest(msgspec.Struct, frozen=True):
    verb: str
    arg: str = ""


class CommandResult(msgspec.Struct, frozen=True):
    """Uniform reply for every mutating endpoint.

    ``ok`` is a claim about what actually happened. A verb that changed nothing
    reports ok=False with a reason — never a cheerful acknowledgement.
    """

    ok: bool
    message: str
    problems: list[str] = msgspec.field(default_factory=list)


class SymbolRequest(msgspec.Struct, frozen=True):
    symbol: str


class SettingsPatch(msgspec.Struct, frozen=True):
    """Partial settings update.

    Both halves arrive as raw mappings, not typed structs: a PUT carrying only
    ``{"app": {"theme": "nord"}}`` must leave every other field alone, and
    decoding straight into AppConfig would silently reset the rest to defaults.
    The route deep-merges onto the CURRENT config before converting.
    """

    app: dict[str, Any] | None = None
    bot: dict[str, Any] | None = None
